### PR TITLE
feat: Using command returns to reply

### LINKED ISSUE #91: feat: Use commands' return value
Commands' return value could be one of:
- `:ok`
- `{:react, emoji}`
- `{:reply, message}`
- `{:private_reply, message}`

Their names are pretty self-explanatory. For example, `{:reply, message}` just maps to `send_message(message, context.channel_id)`

This would be done in the `listen :message_create` generated in `Mobius.Cog.__before_compile__` (or ideally done in a function called from there).

For now we could support only `:ok` and add support for the others in a later PR (or all at once if you feel motivated) since this introduces a breaking change to the command API which should be addressed as soon as possible.

### PR DESCRIPTION
**Pull request by ZeLarpMaster**
Closes #91

This PR adds some testing utilities for cogs. I'm fairly happy with the API for it, though the implementation is a bit shitty because of how Tesla.Mock works. I think that would be worth fixing after the v1.

This PR also includes a fix on the REST client user agent missing a `)`

### CODE REVIEW

File: lib/mobius/cog.ex [hunk 1]
```diff
@@ -290,6 +294,11 @@ defmodule Mobius.Cog do
     end
   end
 
+  @doc false
+  def handle_return(:ok, _context), do: :ok
+  def handle_return({:reply, body}, context), do: send_message(body, context.channel_id)
+  def handle_return(value, _context), do: Logger.warn("Invalid return: #{inspect(value)}")
```
Comment (ZeLarpMaster) [inline lib/mobius/cog.ex]: Had to add this so I could test it. I didn't find a way to detect the cog crashing (I tried `Process.monitor` on the cog, but it crashed without notifying the test :/ ). If you have an idea, I'm open to changing this, I don't really like having to wait for logs.

File: test/support/cog_stub.ex [hunk 2]
```diff
@@ -37,6 +37,14 @@ defmodule Mobius.Stubs.Cog do
     send_to_test({:unexpected_everything, value})
   end
 
+  command "unsupported" do
+    :unsupported_return
+  end
+
+  command "reply" do
+    {:reply, %{content: "The answer"}}
```
Comment (ZeLarpMaster) [inline test/support/cog_stub.ex:45]: Would it be worth changing the other tests which use the cog_stub so they do this too? We wouldn't need the whole `:cog_test_process` thing I think

File: test/support/cog_test_utils.ex [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    mock_global(fn env ->
+      send(test_pid, env)
+      {200, [], %{}}
+    end)
+  end
+
+  @spec assert_message_sent(Message.message_body()) :: :ok
+  def assert_message_sent(message) do
+    json_message = Jason.encode!(message)
+    %Tesla.Env{url: url} = assert_receive %Tesla.Env{body: ^json_message}
```
Comment (ZeLarpMaster) [inline test/support/cog_test_utils.ex]: Should assert it's a `:post`

File: lib/mobius/cog.ex (hunk 1 above)
Comment (Blond11516) [reply lib/mobius/cog.ex]: I agree that doesn't sound ideal, but I have no idea how people usually do this? :thinking: Good enough for now I guess :shrug:

File: test/support/cog_stub.ex (hunk 2 above)
Comment (Blond11516) [reply test/support/cog_stub.ex:45]: Probably in another PR, but yeah I think that would be nice!

File: test/mobius/cog_test.exs [hunk 4]
```diff
@@ -85,6 +86,24 @@ defmodule Mobius.CogTest do
...
+
+      assert_message_sent(%{content: "The answer"})
+    end
+
+    test "should raise an error if the command returns something unsupported" do
+      assert capture_log(fn ->
+               send_command_payload("unsupported")
+               # Give time for the log to appear
+               Process.sleep(10)
+             end) =~ "Invalid return: :unsupported_return"
```
Comment (Blond11516) [inline test/mobius/cog_test.exs]: What if you actually raised? Couldn't you do
```elixir
try do
  send_command_payload("unsupported")
rescue
  e -> assert ...
end
```

**Review by Blond11516 — DISMISSED** [review]

File: test/mobius/cog_test.exs (hunk 4 above)
Comment (ZeLarpMaster) [reply test/mobius/cog_test.exs]: The exception isn't raised in the same process. `send_command_payload` is an indirect `send/2` to the Cog process so we wouldn't be able to get it here. I tried monitoring the process, but that also didn't seem to work. I'll try to investigate that though, I think it'd be useful if we could do that.

Comment (ZeLarpMaster) [reply test/mobius/cog_test.exs]: Figured it out and it seems to work! It looks better though it doesn't assert *why* the cog dies (would be hard and invasive on the implementation details), but it's something.

Writing this made me realize a mixed approach would be even better. We can capture the log and look for the error logged by the unhandled FunctionClauseError and assert on that too 😄

**Review by Blond11516 — APPROVED** [review]

**Pull request merged**